src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -172,23 +173,34 @@ private int[] getFilterIdsArray(final LeafReaderContext context) throws IOExcept
         if (filterWeight == null) {
             return new int[0];
         }
-        final BitSet filteredDocsBitSet = getFilteredDocsBitSet(context, this.filterWeight);
-        final int[] filteredIds = new int[filteredDocsBitSet.cardinality()];
-        int filteredIdsIndex = 0;
+        return bitSetToIntArray(getFilteredDocsBitSet(context, this.filterWeight));
+    }
+
+    private int[] getParentIdsArray(final LeafReaderContext context, final BitSetProducer parentFilter) throws IOException {
+        if (parentFilter == null) {
+            return null;
+        }
+        return bitSetToIntArray(parentFilter.getBitSet(context));
+    }
+
+    private int[] bitSetToIntArray(final BitSet bitSet) {
+        final int[] intArray = new int[bitSet.cardinality()];
+        int index = 0;
         int docId = 0;
-        while (docId < filteredDocsBitSet.length()) {
-            docId = filteredDocsBitSet.nextSetBit(docId);
+        while (docId < bitSet.length()) {
+            docId = bitSet.nextSetBit(docId);
             if (docId == DocIdSetIterator.NO_MORE_DOCS || docId + 1 == DocIdSetIterator.NO_MORE_DOCS) {
```

Comment:
I know this isnt changed in this PR, but this seems wrong. `NO_MORE_DOCS=Integer.MAX_VALUE` ([ref](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java#L140)). So if we are trying to break before the last document, this will not do it.

Reply:
It is correct. From BitSet.java it is mentioned NO_MORE_DOCS is returned if there are no more set bits.
```
  /**
   * Returns the index of the first set bit starting at the index specified. {@link
   * DocIdSetIterator#NO_MORE_DOCS} is returned if there are no more set bits.
   */
```

Reply:
Sorry meant this portion: `docId + 1 == DocIdSetIterator.NO_MORE_DOCS`

Reply:
Yeah. It won't needed.